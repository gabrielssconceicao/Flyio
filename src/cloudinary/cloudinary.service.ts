import { BadRequestException, Injectable } from '@nestjs/common';
import { v2 as cloudinary } from 'cloudinary';
import { Readable } from 'stream';

@Injectable()
export class CloudinaryService {
  constructor() {
    cloudinary.config({
      cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
      api_key: process.env.CLOUDINARY_API_KEY,
      api_secret: process.env.CLOUDINARY_API_SECRET,
    });
  }

  private readonly profileImageFolder = 'profile-pictures';

  private uploadToCloudinary(
    buffer: Buffer,
    uploadConfig: any,
  ): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        // Create a Readable stream from the buffer
        const bufferStream = Readable.from(buffer);

        // Upload config to Cloudinary
        const uploadStream = cloudinary.uploader.upload_stream(
          uploadConfig,
          (error, result) => {
            if (error) {
              reject(
                new BadRequestException(
                  'Error uploading/updating profile picture.',
                ),
              );
            } else {
              // Retornar a URL segura da imagem ap√≥s o upload
              resolve(result.secure_url);
            }
          },
        );

        // Passar o stream para o upload
        bufferStream.pipe(uploadStream);
      } catch {
        reject(
          new BadRequestException('Error uploading/updating profile picture.'),
        );
      }
    });
  }

  private getPublicIdFromUrl(url: string): string {
    const urlParts = url.split('/');
    const filenameWithExtension = urlParts[urlParts.length - 1];

    const filenameParts = filenameWithExtension.split('.');
    const filename = filenameParts[0];

    return filename;
  }

  async uploadProfilePicture(file: Express.Multer.File): Promise<string> {
    const { buffer } = file;

    const renamedFile = `file_${Date.now()}`;

    return this.uploadToCloudinary(buffer, {
      resource_type: 'image',
      folder: 'profile-pictures',
      public_id: renamedFile,
    });
  }

  async updateProfilePicture(
    file: Express.Multer.File,
    prevImgUrl: string,
  ): Promise<string> {
    const { buffer } = file;

    const publicId = this.getPublicIdFromUrl(prevImgUrl);

    return this.uploadToCloudinary(buffer, {
      resource_type: 'image',
      folder: 'profile-pictures',
      public_id: publicId,
      overwrite: true,
    });
  }

  async deleteProfilePicture(profileImgUrl: string) {
    const publicId = this.getPublicIdFromUrl(profileImgUrl);
    return new Promise((resolve, reject) => {
      cloudinary.uploader.destroy(
        `profile-pictures/${publicId}`,
        {
          resource_type: 'image',
          invalidate: true,
        },
        (error) => {
          if (error) {
            reject(new BadRequestException('Error deleting profile picture.'));
          } else {
            resolve(null);
          }
        },
      );
    });
  }

  async uploadPostImage(file: Express.Multer.File): Promise<string> {
    const { mimetype, size, filename } = file;

    // this.checkSize(size, 10 * 1024, 15 * 1024 * 1024);
    // this.validateMimeType(mimetype, ['image/jpeg', 'image/png', 'image/gif']);

    return `Post image uploaded: ${filename}`;
  }
}
