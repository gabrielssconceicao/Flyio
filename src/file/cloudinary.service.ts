import { BadRequestException, Injectable } from '@nestjs/common';
import { v2 as cloudinary } from 'cloudinary';
import { Readable } from 'stream';
@Injectable()
export class CloudinaryService {
  constructor() {
    cloudinary.config({
      cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
      api_key: process.env.CLOUDINARY_API_KEY,
      api_secret: process.env.CLOUDINARY_API_SECRET,
    });
  }

  private renameFile(filename: string): string {
    const extension = filename.split('.').pop();
    const timestamp = Date.now();
    return `file_${timestamp}.${extension}`;
  }

  async uploadProfilePicture(file: Express.Multer.File): Promise<string> {
    const { originalname, buffer } = file;

    // Realizar as verificações de tamanho e tipo de mime

    const renamedFile = this.renameFile(originalname);

    return new Promise((resolve, reject) => {
      try {
        // Create a Readable stream from the buffer
        const bufferStream = Readable.from(buffer);

        // Upload config to Cloudinary
        const uploadStream = cloudinary.uploader.upload_stream(
          {
            resource_type: 'image',
            folder: 'profile-pictures',
            public_id: renamedFile,
          },
          (error, result) => {
            if (error) {
              reject(
                new BadRequestException('Error uploading profile picture.'),
              );
            } else {
              // Retornar a URL segura da imagem após o upload
              resolve(result.secure_url);
            }
          },
        );

        // Passar o stream para o upload
        bufferStream.pipe(uploadStream);
      } catch {
        reject(new BadRequestException('Error uploading profile picture.'));
      }
    });
  }

  async uploadPostImage(file: Express.Multer.File): Promise<string> {
    const { mimetype, size, filename } = file;

    // this.checkSize(size, 10 * 1024, 15 * 1024 * 1024);
    // this.validateMimeType(mimetype, ['image/jpeg', 'image/png', 'image/gif']);

    return `Post image uploaded: ${filename}`;
  }
}
