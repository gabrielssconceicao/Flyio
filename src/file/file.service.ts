import { BadRequestException, Injectable } from '@nestjs/common';

@Injectable()
export class FileService {
  private throwBadRequest(msg: string): void {
    throw new BadRequestException(msg);
  }

  private checkSize(size: number, minSize: number, maxSize: number): void {
    if (size < minSize)
      this.throwBadRequest(
        `File too small. Minimum size: ${minSize / 1024} KB`,
      );
    if (size > maxSize)
      this.throwBadRequest(
        `File too large. Maximum size: ${maxSize / (1024 * 1024)} MB`,
      );
  }

  private validateMimeType(mimetype: string, allowedTypes: string[]): void {
    if (!allowedTypes.includes(mimetype)) {
      this.throwBadRequest(
        `Invalid file type. Allowed types: ${allowedTypes.join(', ')}`,
      );
    }
  }

  async uploadProfilePicture(file: Express.Multer.File): Promise<string> {
    const { mimetype, size, filename } = file;

    this.checkSize(size, 10 * 1024, 5 * 1024 * 1024);
    this.validateMimeType(mimetype, ['image/jpeg', 'image/png']);

    // Simulação de uploa
    return `Profile picture uploaded: ${filename}`;
  }

  async uploadPostImage(file: Express.Multer.File): Promise<string> {
    const { mimetype, size, filename } = file;

    this.checkSize(size, 10 * 1024, 15 * 1024 * 1024);
    this.validateMimeType(mimetype, ['image/jpeg', 'image/png', 'image/gif']);

    return `Post image uploaded: ${filename}`;
  }
}
