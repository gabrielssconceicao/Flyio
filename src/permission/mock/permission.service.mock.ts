export const permissionServiceMock = {
  verifyUserOwnership: jest.fn(),
};
