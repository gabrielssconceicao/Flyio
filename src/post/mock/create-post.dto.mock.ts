import { CreatePostDto } from '../dto';
export const generateCreatePostDtoMock = (): CreatePostDto => {
  return {
    text: 'Hello world',
  };
};
