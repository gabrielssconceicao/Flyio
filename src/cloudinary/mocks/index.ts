import { cloudinaryServiceMock } from './cloudinary.service.mock';
import { generateFileMock, generatedProfilePictureMock } from './file.mock';

export { cloudinaryServiceMock, generateFileMock, generatedProfilePictureMock };
