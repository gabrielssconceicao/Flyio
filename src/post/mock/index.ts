import { postMock } from './post.mock';
import { generateCreatePostDtoMock } from './create-post.dto.mock';
export { postMock, generateCreatePostDtoMock };
