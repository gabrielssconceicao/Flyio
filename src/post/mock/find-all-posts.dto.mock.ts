import { FindAllPostsResponseDto } from '../dto/find-all-posts.dto';
import { generatedPostMock } from './post.mock';

export function generateFindAllPostsDtoMock(): FindAllPostsResponseDto {
  return {
    count: 1,
    items: [generatedPostMock()],
  };
}
