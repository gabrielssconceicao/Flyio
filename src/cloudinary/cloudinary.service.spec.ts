import { Test, TestingModule } from '@nestjs/testing';
import { CloudinaryService } from './cloudinary.service';
import { BadRequestException } from '@nestjs/common';
import { v2 as cloudinary } from 'cloudinary';
import { generateFileMock } from '../mocks/data/file.mock';

jest.mock('cloudinary', () => ({
  v2: {
    config: jest.fn(),
    uploader: {
      upload_stream: jest.fn(),
    },
  },
}));

describe('<CloudinaryService />', () => {
  let service: CloudinaryService;
  let mockFile: Express.Multer.File;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [CloudinaryService],
    }).compile();

    service = module.get<CloudinaryService>(CloudinaryService);
    mockFile = generateFileMock();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('<UploadProfilePicture />', () => {
    it('should upload profile picture and return the secure URL', async () => {
      const mockUrl =
        'https://mocked.cloudinary.com/image/upload/v1/profile-pictures/file_12345.jpg';

      const uploadStreamMock = jest.fn((options, callback) => {
        callback(null, { secure_url: mockUrl });
        return { pipe: jest.fn() };
      });

      cloudinary.uploader.upload_stream = uploadStreamMock as any;

      const result = await service.uploadProfilePicture(mockFile);
      expect(result).toBe(mockUrl);
      expect(uploadStreamMock).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });

    it('should throw BadRequestException on upload error', async () => {
      const uploadStreamMock = jest.fn((options, callback) => {
        callback(new Error('Upload failed'), null);
        return { pipe: jest.fn() };
      });

      cloudinary.uploader.upload_stream = uploadStreamMock as any;

      await expect(service.uploadProfilePicture(mockFile)).rejects.toThrow(
        BadRequestException,
      );
    });
  });
});
