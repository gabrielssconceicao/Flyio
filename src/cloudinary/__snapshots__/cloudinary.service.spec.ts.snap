// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CloudinaryService /> <UploadProfilePicture /> should upload profile picture and return the secure URL 1`] = `"https://mocked.cloudinary.com/image/upload/v1/profile-pictures/file_12345.jpg"`;
